lang/java/avro/src/main/java/org/apache/avro/generic/GenericDatumReader.java

```diff
@@ -257,7 +258,12 @@ protected Object readRecord(Object old, Schema expected, ResolvingDecoder in) th
    */
   protected void readField(Object record, Field field, Object oldDatum, ResolvingDecoder in, Object state)
       throws IOException {
-    data.setField(record, field.name(), field.pos(), read(oldDatum, field.schema(), in), state);
+    try {
+      data.setField(record, field.name(), field.pos(), read(oldDatum, field.schema(), in), state);
+    } catch(AvroTypeException exception) {
+      String message = "Field \"" + field.name() + "\" content mismatch: " + exception.getMessage();
+      throw new AvroTypeException(message);
```

Comment:
In order to not loose the stack trace; use constructor with "cause" parameter; either
`throw new AvroTypeException(message, exception);`
or, to avoid having 2 AvroTypeException
`throw new AvroTypeException(message, exception.getCause());`